File: packages/gh-tasks/src/commands/done.ts
import { resolveLocale, t } from '../i18n/index.ts';
import type { AppConfig } from '../lib/config.ts';
import { createClient, type GraphQLClient, resolveToken } from '../lib/github.ts';
import { ProjectError, type ProjectRef, resolveProjectRef } from '../lib/project.ts';
import { resolveProjectNodeId } from '../lib/projectItem.ts';
import {
  CLOSE_ISSUE,
  type CloseIssueResponse,
  GET_ISSUE_BY_NUMBER,
  type GetIssueByNumberResponse,
  LIST_PROJECT_V2_FIELDS,
  LIST_PROJECT_V2_ITEMS,
  type ListProjectV2FieldsResponse,
  type ListProjectV2ItemsResponse,
  type ProjectV2FieldNode,
  type ProjectV2ItemNode,
  UPDATE_PROJECT_V2_ITEM_FIELD_VALUE,
  type UpdateProjectV2ItemFieldValueResponse,
} from '../lib/queries/index.ts';
import { resolveRepo } from '../lib/repo.ts';
import { detectScope, type Scope } from '../lib/scope.ts';

export interface DoneCommandDeps {
  client?: GraphQLClient;
  hasGitRemote?: () => boolean;
  getRemoteUrl?: () => string | null;
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
  config?: AppConfig;
}

const FIELDS_FETCH_LIMIT = 50;
const ITEMS_FETCH_LIMIT = 100;

export async function done(argv: readonly string[], deps: DoneCommandDeps = {}): Promise<number> {
  const stdout = deps.stdout ?? process.stdout;
  const stderr = deps.stderr ?? process.stderr;
  const locale = resolveLocale(argv, process.env, deps.config);

  const rawId = parsePositionalId(argv);
  if (rawId === null) {
    stderr.write(`${t(locale, 'error.done.idRequired')}\n`);
    return 2;
  }

  const scope = detectScope({ argv, hasGitRemote: deps.hasGitRemote, config: deps.config });

  if (scope === 'repo') {
    return await doneRepoIssue({ argv, deps, locale, stdout, stderr, rawId });
  }

  return await doneProjectItem({ argv, deps, locale, scope, stdout, stderr, rawId });
}

interface DoneRepoContext {
  argv: readonly string[];
  deps: DoneCommandDeps;
  locale: 'ja' | 'en';
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  rawId: string;
}

async function doneRepoIssue(ctx: DoneRepoContext): Promise<number> {
  const { argv, deps, locale, stdout, stderr, rawId } = ctx;

  const issueNumber = toIssueNumber(rawId);
  if (issueNumber === null) {
    stderr.write(`${t(locale, 'error.done.idRequired')}\n`);
    return 2;
  }

  const repo = resolveRepo({ argv, getRemoteUrl: deps.getRemoteUrl });
  const client = deps.client ?? createClient(resolveToken());

  const issueData = await client.request<GetIssueByNumberResponse>(GET_ISSUE_BY_NUMBER, {
    owner: repo.owner,
    name: repo.name,
    number: issueNumber,
  });
  const issue = issueData.repository?.issue;
  if (!issue) {
    stderr.write(`Issue not found: ${repo.owner}/${repo.name}#${issueNumber}\n`);
    return 1;
  }
  if (issue.state === 'CLOSED') {
    stdout.write(`${t(locale, 'done.alreadyClosed')}: ${issue.url}\n`);
    return 0;
  }

  const closed = await client.request<CloseIssueResponse>(CLOSE_ISSUE, {
    input: { issueId: issue.id },
  });
  stdout.write(`${t(locale, 'done.closed')}: ${closed.closeIssue.issue.url}\n`);
  return 0;
}

interface DoneProjectContext {
  argv: readonly string[];
  deps: DoneCommandDeps;
  locale: 'ja' | 'en';
  scope: Exclude<Scope, 'repo'>;
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  rawId: string;
}

async function doneProjectItem(ctx: DoneProjectContext): Promise<number> {
  const { argv, deps, locale, scope, stdout, stderr, rawId } = ctx;

  // For org/user scope `<id>` is the Projects v2 item node id (e.g. `PVTI_xxx`).
  // Looking up a Project item from a repo Issue number is a future extension.
  const itemId = rawId;

  let projectRef: ProjectRef;
  try {
    projectRef = resolveProjectRef({ scope, argv, config: deps.config });
  } catch (err) {
    if (err instanceof ProjectError) {
      stderr.write(`${t(locale, err.i18nKey, err.i18nArgs)}\n`);
      return 2;
    }
    throw err;
  }

  const client = deps.client ?? createClient(resolveToken());
  const projectId = await resolveProjectNodeId({ client, scope, projectRef });
  if (!projectId) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const fieldsData = await client.request<ListProjectV2FieldsResponse>(LIST_PROJECT_V2_FIELDS, {
    projectId,
    first: FIELDS_FETCH_LIMIT,
  });
  if (!fieldsData.node) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const statusField = findStatusField(fieldsData.node.fields.nodes);
  if (!statusField) {
    stderr.write(`${t(locale, 'error.done.statusFieldMissing')}\n`);
    return 1;
  }
  const doneOption = statusField.options.find((o) => o.name.toLowerCase() === 'done');
  if (!doneOption) {
    stderr.write(`${t(locale, 'error.done.doneOptionMissing')}\n`);
    return 1;
  }

  const itemsData = await client.request<ListProjectV2ItemsResponse>(LIST_PROJECT_V2_ITEMS, {
    projectId,
    first: ITEMS_FETCH_LIMIT,
  });
  const targetItem = itemsData.node?.items.nodes.find(
    (item: ProjectV2ItemNode) => item.id === itemId
  );
  if (!targetItem) {
    stderr.write(`item not found in project: ${itemId}\n`);
    return 1;
  }

  if (isAlreadyDone(targetItem, statusField.id, doneOption.id)) {
    stdout.write(`${t(locale, 'done.alreadyDone.project')}: ${itemId}\n`);
    return 0;
  }

  await client.request<UpdateProjectV2ItemFieldValueResponse>(UPDATE_PROJECT_V2_ITEM_FIELD_VALUE, {
    input: {
      projectId,
      itemId,
      fieldId: statusField.id,
      value: { singleSelectOptionId: doneOption.id },
    },
  });
  stdout.write(`${t(locale, 'done.statusUpdated.project')}: ${itemId}\n`);
  return 0;
}

interface SingleSelectStatusField {
  id: string;
  name: string;
  options: Array<{ id: string; name: string }>;
}

function findStatusField(fields: readonly ProjectV2FieldNode[]): SingleSelectStatusField | null {
  for (const f of fields) {
    if (f.dataType === 'SINGLE_SELECT' && f.name.toLowerCase() === 'status') {
      return { id: f.id, name: f.name, options: f.options };
    }
  }
  return null;
}

function isAlreadyDone(
  item: ProjectV2ItemNode,
  statusFieldId: string,
  doneOptionId: string
): boolean {
  for (const v of item.fieldValues.nodes) {
    if (
      v.__typename === 'ProjectV2ItemFieldSingleSelectValue' &&
      v.field.id === statusFieldId &&
      v.optionId === doneOptionId
    ) {
      return true;
    }
  }
  return false;
}

/**
 * Pull the first positional argument from argv, returning it as a raw string.
 * In repo scope this is later coerced to an Issue number; in org/user scope it
 * is treated as a Projects v2 item node id (e.g. `PVTI_xxx`).
 */
function parsePositionalId(argv: readonly string[]): string | null {
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === undefined) continue;
    if (arg.startsWith('--')) {
      // skip value-flags' values
      if (
        !arg.includes('=') &&
        (arg === '--scope' || arg === '--repo' || arg === '--lang' || arg === '--project')
      ) {
        i++;
      }
      continue;
    }
    return arg;
  }
  return null;
}

function toIssueNumber(value: string): number | null {
  const stripped = value.startsWith('#') ? value.slice(1) : value;
  const n = Number.parseInt(stripped, 10);
  return Number.isFinite(n) && n > 0 ? n : null;
}

File: cmd/done.go
package cmd

import (
	"context"
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/spf13/cobra"

	"github.com/ozzy-labs/gh-tasks/internal/github/queries"
	"github.com/ozzy-labs/gh-tasks/internal/project"
	"github.com/ozzy-labs/gh-tasks/internal/projectitem"
	"github.com/ozzy-labs/gh-tasks/internal/repo"
	"github.com/ozzy-labs/gh-tasks/internal/scope"
)

const (
	doneFieldsLimit = 50
	doneItemsLimit  = 100
)

func newDoneCmd(deps Deps) *cobra.Command {
	c := &cobra.Command{
		Use:   "done <id>",
		Short: "Mark a task as done",
		RunE: func(c *cobra.Command, args []string) error {
			return runDone(c.Context(), c, deps, args)
		},
	}
	addJSONFlags(c)
	addJSONCompletion(c, itemJSONFields)
	return c
}

func runDone(ctx context.Context, c *cobra.Command, deps Deps, args []string) error {
	r, err := deps.Resolve(c)
	if err != nil {
		return localizedError(c, r, err)
	}
	jsonReq, jsonOn, err := resolveJSONRequest(c, r, itemJSONFields)
	if err != nil {
		return err
	}
	if len(args) == 0 || args[0] == "" {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.done.idRequired"))
		return ErrSilentArgs
	}
	rawID := args[0]
	sc, err := scope.Detect(scope.DetectOptions{
		Flag:         flagString(c, "scope"),
		HasGitRemote: deps.HasGitRemote,
		DefaultScope: r.Config.DefaultScope,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	if sc == scope.Repo {
		return runDoneRepo(ctx, c, deps, r, rawID, jsonOn, jsonReq)
	}
	return runDoneProject(ctx, c, deps, r, sc, rawID, jsonOn, jsonReq)
}

func runDoneRepo(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, rawID string, jsonOn bool, jsonReq jsonRequest) error {
	num, ok := parseIssueNumber(rawID)
	if !ok {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.done.idRequired"))
		return ErrSilentArgs
	}
	id, err := repo.Resolve(repo.ResolveOptions{Flag: flagString(c, "repo"), GetRemoteURL: deps.GetRemoteURL})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	gqlClient := clients.AsGenqlientClient()
	resp, err := queries.GetIssueByNumber(ctx, gqlClient, id.Owner, id.Name, num)
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "get issue", err)
	}
	if resp.Repository == nil || resp.Repository.Issue == nil {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.issue.notFound", "owner", id.Owner, "name", id.Name, "number", num))
		return ErrSilentRuntime
	}
	if resp.Repository.Issue.State == queries.IssueStateClosed {
		if jsonOn {
			return renderJSONItems(c, r, []map[string]any{{
				"id": resp.Repository.Issue.Id, "number": resp.Repository.Issue.Number,
				"state": "CLOSED", "title": nil, "type": "ISSUE",
				"updatedAt": nil, "url": resp.Repository.Issue.Url,
			}}, jsonReq, itemJSONFields)
		}
		fmt.Fprintf(c.OutOrStdout(), "%s: %s\n", r.T("done.alreadyClosed"), resp.Repository.Issue.Url)
		return nil
	}
	closed, err := queries.CloseIssue(ctx, gqlClient, &queries.CloseIssueInput{IssueId: resp.Repository.Issue.Id})
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "close issue", err)
	}
	if jsonOn {
		return renderJSONItems(c, r, []map[string]any{{
			"id": closed.CloseIssue.Issue.Id, "number": closed.CloseIssue.Issue.Number,
			"state": "CLOSED", "title": nil, "type": "ISSUE",
			"updatedAt": nil, "url": closed.CloseIssue.Issue.Url,
		}}, jsonReq, itemJSONFields)
	}
	fmt.Fprintf(c.OutOrStdout(), "%s: %s\n", r.T("done.closed"), closed.CloseIssue.Issue.Url)
	return nil
}

func runDoneProject(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, sc scope.Scope, itemID string, jsonOn bool, jsonReq jsonRequest) error {
	pref, err := project.Resolve(project.ResolveOptions{
		Scope:       sc,
		Flag:        flagString(c, "project"),
		OrgProject:  r.Config.OrgProject,
		UserProject: r.Config.UserProject,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	pid, err := projectitem.ResolveProjectNodeID(ctx, clients.GraphQL, sc, pref)
	if err != nil {
		return localizedError(c, r, err)
	}
	if pid == "" {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	gqlClient := clients.AsGenqlientClient()
	fieldNodes, err := queries.PaginateProjectV2Fields(ctx, gqlClient, pid, doneFieldsLimit)
	if errors.Is(err, queries.ErrProjectNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list project fields", err)
	}
	fields := projectitem.FieldsOf(fieldNodes)
	statusField := findStatusField(fields)
	if statusField == nil {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.done.statusFieldMissing"))
		return ErrSilentRuntime
	}
	doneOption := findOption(statusField.Options, "done")
	if doneOption == nil {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.done.doneOptionMissing"))
		return ErrSilentRuntime
	}

	itemList, err := queries.PaginateProjectV2Items(ctx, gqlClient, pid, doneItemsLimit)
	if errors.Is(err, queries.ErrProjectNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list project items", err)
	}
	var target *queries.ProjectV2ItemNode
	for _, n := range itemList {
		if n != nil && n.Id == itemID {
			target = n
			break
		}
	}
	if target == nil {
		// Distinguish "not found in the page" from "not found in the
		// project at all": linear search is bounded by doneItemsLimit
		// (= 100) per ADR design — items beyond the limit are reported
		// via searchLimit so the user knows to disambiguate. Resolving
		// items by id directly via Node(id:) is tracked separately and
		// out of scope for the cursor-pagination work.
		if len(itemList) >= doneItemsLimit {
			fmt.Fprintln(c.ErrOrStderr(), r.T("error.done.searchLimit", "id", itemID, "limit", doneItemsLimit))
		} else {
			fmt.Fprintln(c.ErrOrStderr(), r.T("error.projectItem.notFound", "id", itemID))
		}
		return ErrSilentRuntime
	}
	if isAlreadyDone(target, statusField.ID, doneOption.ID) {
		if jsonOn {
			return renderJSONItems(c, r, projectItemRowsToJSON([]*queries.ProjectV2ItemNode{target}), jsonReq, itemJSONFields)
		}
		fmt.Fprintf(c.OutOrStdout(), "%s: %s\n", r.T("done.alreadyDone.project"), itemID)
		return nil
	}
	if _, err := queries.UpdateProjectV2ItemFieldValue(ctx, gqlClient, &queries.UpdateProjectV2ItemFieldValueInput{
		ProjectId: pid,
		ItemId:    itemID,
		FieldId:   statusField.ID,
		Value:     &queries.ProjectV2FieldValue{SingleSelectOptionId: &doneOption.ID},
	}); err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "update item field value", err)
	}
	if jsonOn {
		return renderJSONItems(c, r, projectItemRowsToJSON([]*queries.ProjectV2ItemNode{target}), jsonReq, itemJSONFields)
	}
	fmt.Fprintf(c.OutOrStdout(), "%s: %s\n", r.T("done.statusUpdated.project"), itemID)
	return nil
}

func findStatusField(fields []projectitem.FieldDescriptor) *projectitem.FieldDescriptor {
	for i := range fields {
		f := &fields[i]
		if f.DataType == "SINGLE_SELECT" && strings.EqualFold(f.Name, "status") {
			return f
		}
	}
	return nil
}

func findOption(opts []projectitem.FieldOption, name string) *projectitem.FieldOption {
	for i := range opts {
		if strings.EqualFold(opts[i].Name, name) {
			return &opts[i]
		}
	}
	return nil
}

func isAlreadyDone(item *queries.ProjectV2ItemNode, fieldID, optID string) bool {
	for _, v := range projectitem.FieldValuesOf(item) {
		if v.Typename == "ProjectV2ItemFieldSingleSelectValue" && v.Field.ID == fieldID && v.OptionID == optID {
			return true
		}
	}
	return false
}

func parseIssueNumber(raw string) (int, bool) {
	raw = strings.TrimPrefix(raw, "#")
	n, err := strconv.Atoi(raw)
	if err != nil || n <= 0 {
		return 0, false
	}
	return n, true
}
